### PR TITLE
Refactor some code

### PR DESCRIPTION
**Pull request by takiy33**
Hello, and thank you this project.
I refactor some code.

### CODE REVIEW

**Comment by digitalextremist** [comment]
This looks good @takiy33, thank you.

File: lib/timers/events.rb [hunk 1]
```diff
@@ -60,18 +60,12 @@ def schedule(time, callback)
...
     # Returns the first non-cancelled handle.
     def first
-      while handle = @sequence.last
-        if handle.cancelled?
-          @sequence.pop
-        else
-          return handle
-        end
-      end
+      @sequence.reverse.find { |handle| !handle.cancelled? }
```
Comment (tarcieri) [inline lib/timers/events.rb]: This seems like it both changes the behavior (it's no longer popping the cancelled handles) and the `#reverse` operation is O(n) for n handles.

Did you check how this affected the benchmarks?

Comment (digitalextremist) [inline lib/timers/events.rb]: Four times per snippet:

With change:
```
Serviced 6569 events in 2.008290765 seconds, 3270.9907802740468 e/s.
Serviced 6569 events in 2.01333826 seconds, 3262.795731076875 e/s.
Serviced 6569 events in 2.010146653 seconds, 3267.956919072658 e/s.
Serviced 6569 events in 2.01113889 seconds, 3266.3598670827128 e/s.
```

Without change:

```
Serviced 6569 events in 2.018162446 seconds, 3254.988521254984 e/s.
Serviced 6569 events in 2.010376898 seconds, 3267.577662941764 e/s.
Serviced 6569 events in 2.018099288 seconds, 3255.078317105026 e/s.
Serviced 6569 events in 2.012747453 seconds, 3263.7737523774317 e/s
```

Comment (tarcieri) [inline lib/timers/events.rb]: Do the benchmarks actually test large nubmers of timers?

Comment (digitalextremist) [inline lib/timers/events.rb]: And the cancelled handles would only be popped up to the first uncancelled one. Seems like cleaning `@sequence` truly lumping that into `first` could help, or is the `pop` overhead expected to include the `pop` as a more indirect way of cleaning?

Comment (digitalextremist) [inline lib/timers/events.rb]: @tarcieri added a volume testing benchmark with 20 concurrent timers firing 300 times each:

With change:
```
Serviced 3766885 events in 100.282 seconds, 37562.922558385355 e/s; across 20 timers.
Serviced 3766843 events in 100.192 seconds, 37596.24520919834 e/s; across 20 timers.
```

Without change:
```
Serviced 3766828 events in 100.093 seconds, 37633.281048624776 e/s; across 20 timers.
Serviced 3766936 events in 100.191 seconds, 37597.54868201735 e/s; across 20 timers.
```

The number of tests run is fluxating for some reason, and this machine I'm running it on is highly variable in load, so the tests are there but the results are inconclusive. I will push the benchmark.

Comment (digitalextremist) [inline lib/timers/events.rb]: For now I will revert the old version but keep the new version there, commented out. I'm not as deeply familiar with `Timers` as everything else, especially after the refactor; so I will come back to this when I find myself back in this edge of the Celluverse.

**Comment by ioquatix** [comment]
@digitalextremist this is a completely retarded PR. Why the hell did you merge it?

**Comment by digitalextremist** [comment]
Only one of the methods modified is questionable, because `pop` is removed... but that raises a question as to whether cancelled timers should be cleared by looking for the first active timer. Otherwise it's a small and superficial PR, but a good one.

**Comment by ioquatix** [comment]
> Only one of the methods modified is questionable, because `pop` is removed...

The behaviour of the method is changed entirely and changed from amortised O(1) to O(N). If you don't understand why this is a problem perhaps you shouldn't be modifying code you didn't write without consulting the original author.

> but that raises a question as to whether cancelled timers should be cleared by looking for the first active timer.

Yes, this is an implementation detail and a highly efficient mode of operation. I improved the performance of this code by several orders of magnitude, you'd think I'd know what I'm doing.

> Otherwise it's a small and superficial PR, but a good one.

The other changes are detrimental to the style of the code. This PR is complete BS. As you said, it's superficial, and makes performance WORSE. The fact that you merged it without even testing the implications of the PR is even more shocking.

I don't have a lot of time but I sure as hell hate wasting it on crap like this. Please revert the entire PR and then let's have an OPEN discussion about it before making ANY changes to PERFORMANCE critical code paths.

**Comment by digitalextremist** [comment]
The method was reverted yesterday. Check the code. We can definitely discuss the inclusion of `return` from a style perspective though. I will start a thread on the discussion board.

**Comment by ioquatix** [comment]
Please revert the ENTIRE PR and THEN let's have a discussion about what should be merged.

**Comment by ioquatix** [comment]
@digitalextremist has emailed me outside of this conversation because it appears he has a problem with my conduct. I appreciate any and all feedback, but apart from using a few strong words I stand by everything I said. When I first read it I was incredibly shocked that it had been merged without appropriate discussion. In addition I feel completely disrespected by the ongoing reluctance to revert this PR completely. I believe the correct and respectful procedure for any PR should be to consult the relevant developers BEFORE even considering merging, especially from developers who appear to be contributing for the first time.

I personally believe being brutally honest is critically important to the quality of the code. The alternative, not being honest about bad changes or having critical discussion completely undermine everyone's genuine efforts.

**Comment by digitalextremist** [comment]
If you check your email you will notice the code was reverted, and if you step back from the ledge you will also see a discussion board post on the topic. In general, you can feel free to take it down a few notches.

**Comment by ioquatix** [comment]
@digitalextremist Where is https://github.com/takiy33/timers/commit/16ddb4db3bb8da9a62f81bbe5ca02535af5e5a75 reverted?

**Comment by benlangfeld** [comment]
While I do think that @ioquatix has been excessively aggressive here (perhaps he's let his emotional attachment to this code get the better of him), I do have to agree with the general sentiment that merging this was immature. I too could have hit the merge button, but I recognised that this was something that needed feedback I wasn't qualified to provide. There's a balance to strike between delay and reactionary merging, and this PR leaned toward the latter.

All of that said, no release was cut. A follow-up PR to revert the change can always be issued. `git revert` works very well, and no-one really gets hurt by a little source control mess once in a while. Sure, this merge was inappropriate, but lets point that out calmly, politely request less of a trigger-happy attitude toward that big green button, maybe a quick group hug and, as above, step back from the ledge.

**Comment by digitalextremist** [comment]
Thank you @benlangfeld. On the day, this merge could have waited for more discussion, but in general I am expecting lives do not depend on `master` .. further discussion of standards going into google group.

**Comment by ioquatix** [comment]
> perhaps he's let his emotional attachment to this code get the better of him

Yeah, because that was the real problem here.

**Comment by digitalextremist** [comment]
A real problem would exist if it made it into a release. A real problem is still going on about it when I'm obviously hearing the concern and trying to solve this. So let's keep calm and carry on.

I intentionally emailed you @ioquatix because I don't want to spam people with PR comments, etc. So can we please just let our pulses reach an idle pace and remember we all work together?

**Comment by ioquatix** [comment]
Right, can you please revert https://github.com/takiy33/timers/commit/16ddb4db3bb8da9a62f81bbe5ca02535af5e5a75?

**Comment by digitalextremist** [comment]
@ioquatix I believe I found what you were talking about, I was having trouble finding it because it is differently formatted than previously, but I found that. One moment please, sir.

**Comment by digitalextremist** [comment]
Alright, here is this @ioquatix: 34f6494719e4127672a053018405eb8ff158fc6f

Rather than expanding this into a google group post right now, because honestly I don't want to get into another long-term discussion right now ( and I imagine I'm not the only one ) I will say this:

* I did test the performance of the *other* major change, and found it slightly faster -- but the `pop` behavior was gone, so I still reverted it ( but left it there, commented out )
* The majority of the outcry seemed to me to be regarding using `return` explicitly, which you will see in my "perceived" complete revert, but it was not: it was use of `reverse`
* So with that said, I understand your near heart failure @ioquatix but I ask that you please, please realize that I am not *at all* trying to skip over quality control.

In this case I did make a mistake. The single-line issue escaped my testing eye, and that could have happened whether I merged same day or waited. But I hear @benlangfeld especially *because of how he expressed himself* and I assure both/all of you that I really want to see `1.0` be reached, and I am not at all cutting corners, but I am *extremely* open to people who want to help with our shared goal there. The main thing I saw here was removal of `return` which I supported. Now it seems that is a hot-button issue, and if it is, I want to know why. I hear style concerns, etc -- so we need a decision on what the standard is then. If we have that, and assuming I test thoroughly, we won't have any problems. And please believe, I do a lot of research -- but it's a sheer numbers game here, if I commit a lot and audit a lot, and therefore merge a lot, I am going to make mistakes, but here's the vital point:

### `master` is not a released gem.

That doesn't mean I'm going to be sloppy or hit `master` with things that will be redacted later, regularly. It means I'd rather apologize than ask permission in certain cases, expecting everyone to realize that I respect them a *lot* and I am devoting a lot alongside everyone to see `1.0` get cut.

**Comment by ioquatix** [comment]
I was the one who actually wrote those performance tests and I can tell you that they were not designed to measure this type of scenario - i.e. lots of cancelled timers. Specifically, this case happens in RubyDNS where there are MANY uses of timeout(x) {...} which usually don't timeout. If you want to understand the implications of this PR you won't be able to show it with the existing performance tests. Those tests were designed to push as many events through the system as possible and hit every function many times so that they profiler would show me useful data.

**Comment by digitalextremist** [comment]
I'm not referring to tests you wrote. I tested in different ways, but yes I did not test radical numbers of cancelled tests. I get it @ioquatix. I cannot better understand your issue without letting it go for a bit. Seriously dude. I value your input but the tension level that comes with it is really quite a variable.

:ok_hand: all better.

**Comment by ioquatix** [comment]
If you have additional performance tests perhaps you can add them as specs? It's useful to have some kind of metric to judge changes. It might also make sense to implement a performance test dealing with cancelled timers.

There was some useful thing in MiniTest (?) which allowed you to test an algorithms performance and assert something about it's upper bound - it would run with varying input size N and test that the performance matched expectations (e.g. 1, logN, N, N^2, etc). I don't know if something like that exists for RSpec but it might be nice if we can expect a specific performance characteristic from a given implementation.

**Comment by digitalextremist** [comment]
8854e191d3b36df0602245d6f9523c3ef004c0fc
https://github.com/seattlerb/minitest

**Comment by ioquatix** [comment]
http://madeofcode.com/posts/86-performance-benchmarking-ruby-with-minitest is what I was thinking of.

**Pull request merged**